### PR TITLE
Implement origin concept and browsing contextless documents.

### PR DESCRIPTION
**Pull request by jdm**
These pave the way for implementing other parts of specifications more thoroughly.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8658)

### CODE REVIEW

**Comment by eefriedman** [comment]
The travis build failed... but it looks like it's not related to this PR; somehow travis is doing manifest checking, but homu isn't?

**Comment by jdm** [comment]
Yeah, we haven't updated the builders since https://github.com/servo/saltfs/pull/163 merged.

**Comment by jdm** [comment]
It is related to this PR; I need to run cargo-update again to get rid of a dependency that was added while in development.

**Comment by eefriedman** [comment]
Reviewed 7 of 7 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

_[components/script/dom/document.rs, line 1446 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8658#-K3q3VOROFkfll9PPnv7:-K3q3VOROFkfll9PPnv8:277498194) ([raw file](https://github.com/servo/servo/blob/d47fed4f8174eb9a409606df97631c841b04daee/components/script/dom/document.rs#L1446)):_
Is the meaning of "served over the network" actually defined anywhere in the spec?  If not, please file a spec bug.  ("Has a browsing context" seems like a reasonable approximation for now.)

---

_[components/script/origin.rs, line 9 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8658#-K3q7mLkapAHjs18Hg2h:-K3q7mLkapAHjs18Hg2i:-843482904) ([raw file](https://github.com/servo/servo/blob/d47fed4f8174eb9a409606df97631c841b04daee/components/script/origin.rs#L9)):_
You might want to link to the definition in the HTML spec?  It's not quite the same thing given the inclusion of aliases.

---

_[components/script/origin.rs, line 19 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8658#-K3q7bJP3VtBnkPdsTnn:-K3q7bJP3VtBnkPdsTno:157884004) ([raw file](https://github.com/servo/servo/blob/d47fed4f8174eb9a409606df97631c841b04daee/components/script/origin.rs#L19)):_
This representation of an "alias" isn't really workable; an alias resolves to the origin of another document *even if that origin changes*.

---

_[components/script/origin.rs, line 35 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8658#-K3q5tA2egBo_er_6eX2:-K3q5tA2egBo_er_6eX3:-153479354) ([raw file](https://github.com/servo/servo/blob/d47fed4f8174eb9a409606df97631c841b04daee/components/script/origin.rs#L35)):_
It's probably not a good idea to use the `file:` protocol here; `file:///tmp` isn't a valid URL on Windows.

---

_[components/script/origin.rs, line 71 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8658#-K3q8IVhu7Kmr0Dpj99w:-K3q8IVhu7Kmr0Dpj99x:502485817) ([raw file](https://github.com/servo/servo/blob/d47fed4f8174eb9a409606df97631c841b04daee/components/script/origin.rs#L71)):_
It might be better to force callers to explicitly resolve an Origin to a URLOrigin?  Not sure.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8658)_

**Comment by jdm** [comment]
I've only rebased; I haven't addressed any previous comments.

**Comment by jdm** [comment]
Still only rebased.

**Comment by jdm** [comment]
Review status: 2 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[components/script/dom/document.rs, line 1446 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8658#-K3q3VOROFkfll9PPnv7:-K9nf2Pjzn29X_D3RDjf:-742750599) ([raw file](https://github.com/servo/servo/blob/d47fed4f8174eb9a409606df97631c841b04daee/components/script/dom/document.rs#L1446)):*
Filed https://github.com/whatwg/html/issues/635

---

*[components/script/origin.rs, line 71 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8658#-K3q8IVhu7Kmr0Dpj99w:-K9oaVsZnenfIETQhkoT:-1938217756) ([raw file](https://github.com/servo/servo/blob/d47fed4f8174eb9a409606df97631c841b04daee/components/script/origin.rs#L71)):*
I don't think that makes sense. Origins should be opaque objects that can only be queried for equality and limited properties like the dealiased host.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8658)*

**Comment by jdm** [comment]
All comments should be addressed now.

**Comment by jdm** [comment]
Whoops, I forgot an important piece. I still need to add a lint that forbids `DOMRefCell<Origin>` to close the obvious hole in aliasing.

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -2318,11 +2381,14 @@ impl DocumentMethods for Document {
 
     // https://html.spec.whatwg.org/multipage/#dom-document-cookie
     fn GetCookie(&self) -> Fallible<DOMString> {
-        // TODO: return empty string for cookie-averse Document
-        let url = self.url();
-        if !is_scheme_host_port_tuple(&url) {
+        if self.is_cookie_averse() {
+            return Ok(DOMString::from("".to_owned()));
```
Comment (frewsxcv) [inline components/script/dom/document.rs]: This could be just `DOMString::new()`

**Comment by KiChjang** [comment]
```

Ran 4566 tests finished in 229.0 seconds.
  • 4564 ran as expected. 1371 tests skipped.
  • 1 tests failed unexpectedly
  • 1 tests had unexpected subtest results

Tests with unexpected results:
  ▶ Unexpected subtest result in /html/browsers/origin/relaxing-the-same-origin-restriction/document_domain.html:
  └ PASS [expected FAIL] new document

  ▶ FAIL [expected PASS] /_mozilla/css/mix_blend_mode_a.html
  └   → /_mozilla/css/mix_blend_mode_a.html d9d9e489bc5b4508c01b06c7ff92f19dcc2ece3c
/_mozilla/css/mix_blend_mode_ref.html cab28df3ff5ab6f04e34b81019bde496e05d8965
Testing d9d9e489bc5b4508c01b06c7ff92f19dcc2ece3c == cab28df3ff5ab6f04e34b81019bde496e05d8965
```
Looking good here! Except for the intermittent...

**Comment by ecoal95** [comment]
See #9779

**Comment by jdm** [comment]
This will require changes to address https://github.com/whatwg/html/pull/868.

**Comment by Ms2ger** [comment]
-S-needs-rebase 

LGTM.

r? @jdm to check my changes.

---

Reviewed 4 of 15 files at r3, 4 of 14 files at r6, 4 of 14 files at r7, 1 of 3 files at r8, 3 of 8 files at r10, 1 of 1 files at r11, 1 of 3 files at r13, 1 of 3 files at r19, 1 of 9 files at r21, 5 of 7 files at r24, 3 of 3 files at r25, 5 of 5 files at r26, 3 of 3 files at r27.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/8658#-:-KF8yCbcLwG0DEgaHWvv:430526668)*

**Comment by jdm** [comment]
This looks good to me!

**Pull request merged**